File: eecs268/2021spring/MWF/2021.04.21.BST-remove/BST.cpp
//public search
template <typename ItemType, typename KeyType>
ItemType BST<ItemType, KeyType>::search(KeyType key) const
{
	return( search(key, m_root) );
}

//private
template <typename ItemType, typename KeyType>
ItemType BST<ItemType, KeyType>::search(KeyType key, BNode<ItemType>* subtree) const
{
	//Board work: finish this definition
	//Assume that ItemType == KeyType
	//				ItemType < KeyType
	//				ItemType > KeyType
	//			are all defined
	if(subtree == nullptr)
	{
		//throw an exception
	}
	else if(subtree->getEntry() == key)
	{
		return(subtree->getEntry());
	}
	else if(subtree->getEntry() < key)
	{
		//go right
	}
	else
	{
		//go left
	}
}

//public facing add
template <typename ItemType, typename KeyType>
void BST<ItemType, KeyType>::add(ItemType entry)
{
	//Deal with the case of adding to empty tree
	if(m_root == nullptr)
	{
		m_root = new BNode<ItemType>(entry);		
	}
	else
	{
		recAdd(entry, m_root);
	}
}

//private recursive add
template <typename ItemType, typename KeyType>
void BST<ItemType, KeyType>::recAdd(ItemType entry, BNode<ItemType>* subtree)
{
	//subtree pointer will ALWAYS be pointing to a BNode object
	
	//Base case 1: entry matches the current node's entry
	//Base case 2&3: The entry should be added to LST/RST and that subtree is empty
	//				  then just add the new entry
	//Recursive case: Go into appropriate subtree
}

File: eecs268/2024spring/TR/2024.03.19.bst-intro/bst.py
#bst.py
from binarynode import BinaryNode
class BST:
    def __init__(self):
        self._root = None

    def add(self, entry):
        #public facing, non-recursive
        if self._root is None:
            self._root = BinaryNode(entry)
        else:
            self._rec_add(entry, self._root)

    def _rec_add(self, entry, cur_node):
        #hidden, recursive
        if entry < cur_node.entry:
            if cur_node.left is None:
                #add it!
            else:
                self._rec_add(entry, cur_node.left)
        #finish this definition for board work
